mlir-sycl/doc/design/SYCLKernelSubmissionRepresentation.md

```diff
@@ -0,0 +1,528 @@
+
+At this point, the handler operand is merely an implementation detail which
+prevents the operation from being considered trivially dead without interfering
+with the dataflow analyses defined in context of the SYCL dialect. Also note
+that we do not model events or dependencies between different kernel launches
+yet — a future extension of the operation may have a result and operands
+representing events.
+
+## Raising approach
+
+Lots of high-level structure and information is lost during the lowering of the
+object-oriented SYCL source code to LLVM-IR. In order to recover the information
+required for the `schedule_kernel` operation, we employ a multi-stage pattern
+matching approach that is aided by source code annotations in the SYCL headers.
+The following subsections describe the intermediate steps in detail. The
+implementation resides in
+`polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp`. Note that
+example MLIR snippets have been manually edited to use descriptive names for
+values and other identifiers.
+
+### Command-group handler
+
+All raising patterns operate on the CGF itself or annotation known to reside in
+the CGF after mandatory inlining (see
+[above](#prerequisites--known-limitations)). Hence, the handler is the first
+argument (after the `this` pointer) of the CGF.
+
+### Host constructors
+
+We raise calls/invokes to specific constructors of SYCL API classes (`buffer`,
+`accessor`, `range`, etc.) to the
+[`sycl.host.constructor`](#syclhostconstructor-syclsyclhostconstructorop) op,
+which ties together arguments for construction and the SYCL dialect type of the
+entity being initialized.
+
+For example, the instantiation of an accessor in the CGF results in the
+following raised IR:
+```mlir
+%accessor = llvm.alloca %c1 x !llvm.struct<"class.sycl::_V1::accessor", // ...
+...
+sycl.host.constructor(%accessor, %buffer, %handler, %propertylist, %codeloc)
+  {type = !sycl_accessor_1_21llvm2Evoid_r_gb}
+    : (!llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr, !llvm.ptr) -> ()
+```
+
+The host-side constructor operation differs from the `sycl.constructor`
+operation used in device code in the following aspects:
+- `sycl.host.constructor` does *not* include the allocation of the object.
+  Hence, a common pattern in the host raising process is looking for a
+  `host.constructor` op in the users of a particular `llvm.alloca` operation.
+- As SYCL objects are passed around as `!llvm.ptr` values in the host IR
+  (instead of `memref`s of SYCL dialect types), the operation carries a type
+  attribute to encode additional high-level information (see below) about the
+  constructed SYCL entities.
+
+We currently extract the following information and encode it in the type
+attribute:
+- `sycl::buffer`: dimensions, sub-buffer flag
+- `sycl::accessor`: dimensions, access mode, access target, presence of a
+  (sub-)range and offset
+- `sycl::range`, `sycl::id`, `sycl::nd_range`: dimensions
```

Comment:
@sommerlukas knows more about this, but I'd add that we do not add additional fields to `accessor`, but encode extra info (range being passed to the constructor and offset) in the body (can be `void`, `range` or `range+id` IIRC)

Reply:
Yes, probably worth mentioning that here. As Victor said, it's encoded via the body type list.

Reply:
Done.